### PR TITLE
Add loongarch64 support to bridges

### PR DESCRIPTION
**Pull request by wjh-la**

### CODE REVIEW

File: solenv/clang-format/excludelist [hunk 1]
```diff
@@ -448,6 +448,10 @@ bridges/source/cpp_uno/gcc3_linux_aarch64/callvirtualfunction.cxx
 bridges/source/cpp_uno/gcc3_linux_aarch64/callvirtualfunction.hxx
 bridges/source/cpp_uno/gcc3_linux_aarch64/cpp2uno.cxx
 bridges/source/cpp_uno/gcc3_linux_aarch64/uno2cpp.cxx
+bridges/source/cpp_uno/gcc3_linux_loongarch64/abi.cxx
+bridges/source/cpp_uno/gcc3_linux_loongarch64/abi.hxx
+bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx
+bridges/source/cpp_uno/gcc3_linux_loongarch64/uno2cpp.cxx
```
Comment (xen0n) [inline solenv/clang-format/excludelist]: The list is sorted alphabetically, you should add the lines somewhere below.

File: bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+    reinterpret_cast<unsigned int *>(code)[4] = 0x14000014
+    	| (((((unsigned long)vtableSlotCall) >> 12) & 0x000fffff) << 5);
+    reinterpret_cast<unsigned int *>(code)[5] = 0x03800294
+	| ((((unsigned long)vtableSlotCall) & 0x00000fff) << 10);
+    reinterpret_cast<unsigned int *>(code)[6] = 0x16000014
+	| (((((unsigned long)vtableSlotCall) >> 32) & 0x000fffff) << 5);
+    reinterpret_cast<unsigned int *>(code)[7] = 0x03000294
+	| (((((unsigned long)vtableSlotCall) >> 52) & 0x00000fff) << 10);
+    // jump $t8
+    reinterpret_cast<unsigned int *>(code)[8] = 0x4c000280;
```
Comment (xen0n) [inline bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx]: mention the assembly in comments for these insns, similar to the 4 insns above?

File: bridges/source/cpp_uno/gcc3_linux_loongarch64/uno2cpp.cxx [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+        // Fill the general purpose registers
+        "ld.d $r4, %[gpr], 0\n\t"
+        "ld.d $r5, %[gpr], 8 \n\t"
+        "ld.d $r6, %[gpr], 16\n\t"
+        "ld.d $r7, %[gpr], 24\n\t"
+        "ld.d $r8, %[gpr], 32\n\t"
+        "ld.d $r9, %[gpr], 40\n\t"
+        "ld.d $r10, %[gpr], 48\n\t"
+        "ld.d $r11, %[gpr], 56\n\t"
+        // Fill the floating pointer registers
```
Comment (xen0n) [inline bridges/source/cpp_uno/gcc3_linux_loongarch64/uno2cpp.cxx]:
```suggestion
        // Fill the floating-point registers
```

File: bridges/source/cpp_uno/gcc3_linux_loongarch64/uno2cpp.cxx [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+        "fld.d $f5, %[fpr], 40\n\t"
+        "fld.d $f6, %[fpr], 48\n\t"
+        "fld.d $f7, %[fpr], 56\n\t"
+        // Perform the call
+        "jirl $ra, %[mfunc], 0\n\t"
+        // Fill the return values
+        "move %[gret1], $r4 \n\t"
+        "move %[gret2], $r5 \n\t"
+        "fmov.d %[fret1], $f0 \n\t"
+        "fmov.d %[fret2], $f1 \n\t"
```
Comment (xen0n) [inline bridges/source/cpp_uno/gcc3_linux_loongarch64/uno2cpp.cxx]: remove the stray space before `\n\t`

File: bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+    }
+}
+
+struct loongarch64_va_list {
+    void * stack;
+    void * gr_top;
+    void * vr_top;
+    int gr_offs;
+    int vr_offs;
+};
```
Comment (xen0n) [inline bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx]: I have verified this is wrong. The [LoongArch gcc port](https://github.com/gcc-mirror/gcc/blob/2aca9d5b92723ee3e34b011902582e8738208695/gcc/config/loongarch/loongarch-builtins.cc#L420-L424) defines the `va_list` to be just a pointer, while you copied from the AArch64 port.

File: bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx (hunk 2 above)
Comment (xen0n) [reply bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx]: Actually these are the `lu12i.w + ori + lu32i.d + lu52i.d` plus `jr $t8`.

File: bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+    return code;
+}
+
+void bridges::cpp_uno::shared::VtableFactory::flushCode(
+    unsigned char const * begin, unsigned char const * end)
+{
+   static void (*clear_cache)(unsigned char const *, unsigned char const *)
+       = (void (*)(unsigned char const *, unsigned char const *)) dlsym(
+           RTLD_DEFAULT, "__clear_cache");
+   (*clear_cache)(begin, end);
```
Comment (xen0n) [inline bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx]: This is again wrong. `__clear_cache` is no-op on LoongArch, as the LoongArch gcc doesn't define `CLEAR_INSN_CACHE` for libgcc.

A simple `ibar 0` should be enough.

File: solenv/clang-format/excludelist (hunk 1 above)
Comment (xen0n) [reply solenv/clang-format/excludelist]: Actually, new code *should* get properly formatted.

During local testing and tweaking of this code, I noticed an awful lot of style inconsistencies, so you'd be better off just `clang-format`ting all of the files.

**Review by xen0n — CHANGES_REQUESTED** [review]

File: bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx (hunk 2 above)
Comment (wjh-la) [reply bridges/source/cpp_uno/gcc3_linux_loongarch64/cpp2uno.cxx]: I have revised it again

**Comment by xen0n** [comment]
Still not working.

```
$ ./workdir/CustomTarget/testtools/bridgetest/bridgetest_server 

> accepting socket,host=127.0.0.1,port=2002...connection established.warn:sal.osl:1145227:1145278:sal/osl/unx/socket.cxx:1570: receive socket [0] failed: EOL
warn:binaryurp:1145227:1145278:binaryurp/source/reader.cxx:124: caught UNO exception 'N3com3sun4star3uno9ExceptionE msg: acc_socket.cxx:SocketConnection::read: error - Success at /home/xenon/src/libreoffice/core-rel/io/source/acceptor/acc_socket.cxx:175'
```

```
$ ./workdir/CustomTarget/testtools/bridgetest/bridgetest_client 
./workdir/CustomTarget/testtools/bridgetest/bridgetest_client: line 1: 1145276 Segmentation fault      (core dumped) /home/xenon/src/libreoffice/core-rel/instdir/program/uno.bin -s com.sun.star.test.bridge.BridgeTest -- -u 'uno:socket,host=127.0.0.1,port=2002;urp;test' -env:LO_BUILD_LIB_DIR=file:///home/xenon/src/libreoffice/core-rel/workdir/LinkTarget/Library -env:URE_MORE_SERVICES=file:///home/xenon/src/libreoffice/core-rel/workdir/Rdb/uno_services.rdb -env:URE_MORE_TYPES=file:///home/xenon/src/libreoffice/core-rel/workdir/UnoApiTarget/bridgetest.rdb
```

And of course the build doesn't work (no window shown after splash progress reaching 100%).

**Comment by xen0n** [comment]
Hmm, it actually works, I was using remote X11 so the UI is excessively slow but eventually it appeared.

Better fix the bridge test before upstreaming though...

**Pull request merged**